Tidbit/Generator/Categories.php

```diff
@@ -0,0 +1,191 @@
+                $category = new stdClass();
+                $category->lvl = $i;
+                $categoryByLevels[$i][] = $category;
+                $parent = $categoryByLevels[$i - 1][0];
+                $parent->children[] = $category;
+            }
+        }
+
+        $commonCounter = 0;
+        $this->countTree($categoryByLevels[0], $commonCounter);
+
+        $rootId = $this->createInsertRecord($root);
+        for ($i = 1; $i < $this->nestingLevel; $i++) {
+            foreach ($categoryByLevels[$i] as $category) {
+                $this->createInsertRecord($category, $rootId);
+            }
+        }
+
+        processQueries($this->insertObject->getHead(), $this->insertObject->getValues());
+        $this->insertCounter += count($this->insertObject->getValues());
+        $this->setCategoryRootInConfig($rootId);
+    }
+
+    /**
+     * Remove generated data from DB.
+     */
+    public function clearDB()
+    {
+        $this->db->query("DELETE FROM categories WHERE id LIKE 'seed-%'");
+    }
+
+    /**
+     * Remove all data from the tables of DB affected by generator.
+     */
+    public function obliterateDB()
+    {
+        $this->db->query("DELETE FROM categories");
+    }
+
+    /**
+     * Count tree indexes.
+     *
+     * @param array $categoryByLevels
+     * @param int $commonCounter
+     */
+    private function countTree(array $categoryByLevels, &$commonCounter)
+    {
+        foreach ($categoryByLevels as $category) {
+            $category->lft = ++$commonCounter;
+            if (!empty($category->children)) {
+                $this->countTree($category->children, $commonCounter);
+            }
+            $category->rft = ++$commonCounter;
+        }
+    }
+
+    /**
+     * Add insert record and return id of it.
+     *
+     * @param stdClass $category
+     * @param string $rootId
+     */
+    private function createInsertRecord(stdClass $category, $rootId = '') {
+        $dataTool = $this->getDataToolForModel(static::MODEL_NAME, $this->modelCounter++);
+        $dataTool->installData['root'] = $rootId ? $rootId : $dataTool->installData['id'];
+        $dataTool->installData['lft'] = $category->lft;
+        $dataTool->installData['rgt'] = $category->rft;
+        $dataTool->installData['lvl'] = $category->lvl;
+
+        if (!$this->insertObject) {
+            $this->insertObject = new Tidbit_Generator_Insert_Object(
+                $dataTool->createInsertHead($dataTool->table_name),
+                array($dataTool->createInsertBody())
+            );
+        } else {
+            $this->insertObject->addValues($dataTool->createInsertBody());
+        }
+
+        return $dataTool->installData['id'];
+    }
+
+    /**
+     * Set KB categories root in config table
+     *
+     * @param $rootId
+     */
+    private function setCategoryRootInConfig($rootId)
+    {
+        $this->db->query("UPDATE config SET value={$rootId} WHERE category='KBContents' AND name='category_root'");
```

Comment:
I guess it's good practice to quote all data in SQL, you receive `$rootId` as function parameter so it can be invalid...